clippy_lints/src/copies.rs

```diff
@@ -103,47 +113,457 @@ declare_clippy_lint! {
+
+        let cond_span = first_line_of_span(cx, if_expr.span).until(block.span);
+        let cond_snippet = reindent_multiline(snippet(cx, cond_span, "_"), false, None);
+        let cond_indent = indent_of(cx, cond_span);
+        let moved_span = block.stmts[0].span.source_callsite().to(span_end);
+        let moved_snippet = reindent_multiline(snippet(cx, moved_span, "_"), true, None);
+        let suggestion = moved_snippet.to_string() + "\n" + &cond_snippet + "{";
+        let suggestion = reindent_multiline(Cow::Borrowed(&suggestion), true, cond_indent);
+
+        let span = span_start.to(span_end);
+        suggestions.push(("start", span, suggestion.to_string()));
+    }
+
+    if lint_end {
+        let block = blocks[blocks.len() - 1];
+        let span_end = block.span.shrink_to_hi();
+
+        let moved_start = if end_stmts == 0 && block.expr.is_some() {
+            block.expr.unwrap().span
+        } else {
+            block.stmts[block.stmts.len() - end_stmts].span
+        }
+        .source_callsite();
+        let moved_end = block
+            .expr
+            .map_or_else(|| block.stmts[block.stmts.len() - 1].span, |expr| expr.span)
+            .source_callsite();
+
+        let moved_span = moved_start.to(moved_end);
+        let moved_snipped = reindent_multiline(snippet(cx, moved_span, "_"), true, None);
+        let indent = indent_of(cx, if_expr.span.shrink_to_hi());
+        let suggestion = "}\n".to_string() + &moved_snipped;
+        let suggestion = reindent_multiline(Cow::Borrowed(&suggestion), true, indent);
+
+        let mut span = moved_start.to(span_end);
+        // Improve formatting if the inner block has indention (i.e. normal Rust formatting)
+        let test_span = Span::new(span.lo() - BytePos(4), span.lo(), span.ctxt());
+        if snippet_opt(cx, test_span)
+            .map(|snip| snip == "    ")
+            .unwrap_or_default()
+        {
+            span = span.with_lo(test_span.lo());
+        }
+
+        suggestions.push(("end", span, suggestion.to_string()));
+        add_expr_note = !cx.typeck_results().expr_ty(if_expr).is_unit()
+    }
+
+    let add_optional_msgs = |diag: &mut DiagnosticBuilder<'_>| {
+        if add_expr_note {
+            diag.note("The end suggestion probably needs some adjustments to use the expression result correctly");
+        }
+
+        if warn_about_moved_symbol {
+            diag.warn("Some moved values might need to be renamed to avoid wrong references");
+        }
+    };
+
+    // Emit lint
+    if suggestions.len() == 1 {
+        let (place_str, span, sugg) = suggestions.pop().unwrap();
+        let msg = format!("all if blocks contain the same code at the {}", place_str);
+        let help = format!("consider moving the {} statements out like this", place_str);
+        span_lint_and_then(cx, SHARED_CODE_IN_IF_BLOCKS, span, msg.as_str(), |diag| {
+            diag.span_suggestion(span, help.as_str(), sugg, Applicability::Unspecified);
```

Comment:
Since we know that the suggestions are not perfect, `MaybeIncorrect` seems like the better choice for all the lint suggestions

```suggestion
            diag.span_suggestion(span, help.as_str(), sugg, Applicability::MaybeIncorrect);
```

Reply:
The `Applicability` was set to `Unspecified` on purpose. 

The lint at first added suggestions to each block which made it applicable but the report quite unreadable. @_flip1995 then mentioned that rustfix doesn't work for multiple spans. I therefore decided to limit the lint suggestion to the bottom and the top statements of the bottom and top blocks to make the suggestion readable. The suggestion is therefore not applicable anymore.

An example, the code:

```rust
if true {
    let start = 0x44;
    let middle = "Separator 1";
    let end = 0x99;
} else {
    let start = 0x44;
    let middle = "Separator 2";
    let end = 0x99;
}
```

would cause only these suggestions if they were applied:
```rust
let start = 0x44;
if true {
    let middle = "Separator 1";
    let end = 0x99; // The removal is not suggested it would otherwise just show a
                    // blank line and dilute the actual message focus. The actual move of
                    // these statement is suggested on the end block on the end statements
} else {
    let start = 0x44;  // This removal is not suggested. See above
    let middle = "Separator 2";
}
let end = 0x99;
```

This removes the ability to use rust fix but enables us to have a clean report like this:

```
error: all if blocks contain the same code at the start and the end. Here at the start
  --> $DIR/some_rust_file.rs:16:5
   |
LL | /     if true {
LL | |         let start = 0x44;
   | |_________________________^
   |
note: and here at the end
  --> $DIR/some_rust_file.rs:28:5
   |
LL | /         let end = 0x99;
LL | |     }
   | |_____^
help: consider moving the start statements out like this
   |
LL |     let start = 0x44;
LL |     if x == 7 {
   |
help: and consider moving the end statements out like this
   |
LL |     }
LL |     let end = 0x99;
   |
```

I hope this makes sense and is also okay 🙃

Reply:
Seem fine, thanks for the explanation!